### PR TITLE
[WIP] PoC adding integration tests to purchases module

### PR DESCRIPTION
**Pull request by tonidero**
### Description
Working on a PoC of adding integration tests to the `purchases` module. This allows us to access internal APIs and inject a custom billing wrapper to be able to mock the stores. This is still WIP and hacky, but shows a possible approach to adding tests that mock the store to be able to have more comprehensive test scenarios.

Basically I'm adding a new flavor to the `purchases` module called `integrationTest`. Since it's part of the `purchases` module, we can access internal classes, like the `PurchasesFactory` to allow injection of a mock `BillingAbstract` which we can use to test the rest of the SDK bypassing Google/Amazon.

Tests can be run by changing `REVENUECAT_API_KEY` with your key and running `./gradlew :purchases:connectedCheck` in the console from the root folder or by switching the purchases module variant to the `integrationTestDebug` variant 
image
 And then running `PurchaseIntegrationTest.kt`.
image

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -2,14 +2,14 @@ buildscript {
     ext.kotlinVersion = "1.6.21"
     ext.compileVersion = 31
     ext.buildToolsVersion = "30.0.3"
-    ext.minVersion = 14
+    ext.minVersion = 21
```
Comment (tonidero) [inline build.gradle:5]: This was needed due to multidex support + mockk-android needing android 21+. I still need to see if we can split the module min version for different buildTypes.

File: purchases/src/integrationTest/AndroidManifest.xml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+    <application
+            android:allowBackup="true"
+            android:icon="@mipmap/ic_launcher"
+            android:label="@string/app_name"
+            android:roundIcon="@mipmap/ic_launcher_round"
+            android:supportsRtl="true"
+            android:theme="@style/Theme.Purchasesandroid">
+        <activity android:name="com.revenuecat.purchases.MainActivity" android:exported="true"/>
+    </application>
```
Comment (tonidero) [inline purchases/src/integrationTest/AndroidManifest.xml:12]: Most of this is copied from the `integrationTest` project

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/PurchaseIntegrationTest.kt [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        scenario.onActivity {
+            Purchases.sharedInstance.getCustomerInfoWith({
+                fail("should be success. Error: ${it.message}")
+            }) {
+                lock.countDown()
+            }
+        }
+        lock.await(testTimeout.inWholeSeconds, TimeUnit.SECONDS)
+        assertThat(lock.count).isZero
+    }
```
Comment (tonidero) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/PurchaseIntegrationTest.kt:75]: These tests were copied from the `integrationTests` module

File: purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesFactory.kt [hunk 4]
```diff
@@ -37,7 +37,8 @@ internal class PurchasesFactory(
     fun createPurchases(
         configuration: PurchasesConfiguration,
         platformInfo: PlatformInfo,
-        proxyURL: URL?
+        proxyURL: URL?,
+        overrideBillingAbstract: BillingAbstract? = null
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesFactory.kt:41]: This is a hacky approach to allow injecting a mock billing abstract for now. We might consider better approaches, but wanted to check thoughts from people before.

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/purchasesExtensions.kt [hunk 5]
```diff
@@ -0,0 +1,19 @@
+package com.revenuecat.purchases
+
+import android.annotation.SuppressLint
+import com.revenuecat.purchases.common.BillingAbstract
+
+fun Purchases.Companion.configure(
+    configuration: PurchasesConfiguration,
+    overrideBillingAbstract: BillingAbstract
+): Purchases {
```
Comment (tonidero) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/purchasesExtensions.kt:9]: This method is only available for the new flavor.

**Comment by tonidero** [comment]
This requires much more cleanup. But I think we can see the approach I'm taking here and I thought I would share it to gather feedback @RevenueCat/coresdk.

There are other approaches we could take though. If we decide to make a public API to override Google/Amazon this won't be needed. We can discuss approaches next week!

**Comment by tonidero** [comment]
@vegaro Any thoughts or suggestions on this approach? I think it allows us to add integration tests for `getOfferings` and making purchases (we still would need to add a token, which can be done as part of the CI process). 

Making sure this approach is ok with everyone before moving forward with it.

**Comment by vegaro** [comment]
I think this makes sense @tonidero and it opens room for testing a lot of scenarios we don't have tests for

**Comment by tonidero** [comment]
Closing this in favor of #899

**Pull request closed without merging**